proposals/0000-implicit-resource-binding.md

```diff
@@ -0,0 +1,369 @@
+
+Since DXC seems to only assign descriptors from `space0`, it reports an error
+whenever there are:
+- two or more unbound resource arrays without explicit binding
+- one unbound resource array with explicit binding in `space0` and one or more
+  unbound resource arrays without explicit binding
+
+#### Example 3.4
+
+```c++
+RWBuffer<float> C[];  // gets u0 (unbounded range)
+RWBuffer<float> D[];  // error
+
+[numthreads(4,1,1)]
+void main() {
+  D[10][0] = C[20][2];
+}
+```
+https://godbolt.org/z/od7o6xfWE
+
+Unbound resource arrays cannot be placed inside structs and DXC reports `error:
+array dimensions of struct/class members must be explicit`.
+
+### Register spaces other than `space0`
+
+Use of other virtual register spaces does not seem to affect the implicit
+assignments in `space0`:
+
+#### Example 4.1
+```c++
+....
+RWBuffer<float> A : register(u0, space1); 
+RWBuffer<float> B[10];   // gets u0 in space0 (range 10)
+RWBuffer<float> C[5][4] : register(u10, space5);
+RWBuffer<float> D;       // gets u10 in space0
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[0] + C[3][3][0] + B[5][0];
+}
+```
+https://godbolt.org/z/EeW17MncE
+
+### Space-only register annotations
+
+DXC support binding annotations that only specify the register space from which the descriptor binding should be assigned.
+
+#### Example 5.1
+
+```c++
+RWBuffer<float> A : register(u1);        // defaults to space0
+RWBuffer<float> B[];                     // gets u2 (unbounded range)
+RWBuffer<float> C[3] : register(space1); // gets u0, space1 (range 3)
+RWBuffer<float> D : register(space1);    // gets u3, space1
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = C[2][0] + D[0] + B[10][0];
+}
+```
+https://godbolt.org/z/3Kc1T3dvs
+
+### Unused resources
+
+Unused resources are resources that are not referenced in the source code or
+that are unreachable from the shader entry point function and later removed from
+the final IR code by LLVM optimization passes.
+
+When a resource is unused and it has an explicit binding, this binding is
+ignored and its descriptor range is available for use by other resources.
+
+In the following example the resources `B`, `C` and `s.D` are identified as
+unused and optimized away. Resources `A` and `E` get the binding from descriptor
+range originally assigned to `B`:
+
+#### Example 6.1
+
+```c++
+struct S {
+  RWBuffer<float> D;    
+};
+
+RWBuffer<float> A; // gets u0
+RWBuffer<float> B[10] : register(u0);  // unused
+RWBuffer<float> C : register(u2);      // unused
+S s : register(u5);                    // unused
+RWBuffer<float> E[5]; // gets u1
+
+void foo() {
+    B[2][0] = 1.0;
+}
+
+bool bar(int i) {
+    return i > 10;
+}
+
+[numthreads(4,1,1)]
+void main(uint3 gi : SV_GROUPID) {
+  if (false) {
+    C[0] = 1.0;
+  }
+  for (int i = 0; i < 5; i++) {
+    if (i > 6)
+      s.D[0] = 1.0; // unreachable
+    if (bar(i))
+      foo();        // unreachable
+    if (bar(gi.x))
+      A[0] = E[2][i];
+  }
+}
+```
```

Comment:
Could you add a newline after? Strange how the text is blue here.

Reply:
I am not sure what you mean. The blue link to compiler explorer?

Reply:
All the text under the tick marks is blue, including the link, and `Based on resources ident...`